masonry/src/render_root.rs

```diff
@@ -497,44 +453,22 @@ impl RenderRoot {
         kurbo::Size::new(size.width, size.height)
     }
 
-    // --- MARK: POST-EVENT ---
-    fn post_event_processing(&mut self, widget_state: &mut WidgetState) {
-        // If children are changed during the handling of an event,
-        // we need to send RouteWidgetAdded now, so that they are ready for update/layout.
-        if widget_state.children_changed {
-            run_update_new_widgets_pass(self);
-        }
-
-        if self.state.debug_logger.layout_tree.root.is_none() {
-            self.state.debug_logger.layout_tree.root = Some(self.root.id().to_raw() as u32);
-        }
-
-        if !self.state.scroll_request_targets.is_empty() {
-            run_update_scroll_pass(self);
-        }
-
-        if self.root_state().needs_compose && !self.root_state().needs_layout {
-            root_compose(self, widget_state);
-        }
-
-        // Update the disabled and stashed state if necessary
-        // Always do this before updating the focus-chain
-        if self.root_state().needs_update_disabled {
-            run_update_disabled_pass(self);
-        }
-        if self.root_state().needs_update_stashed {
-            run_update_stashed_pass(self);
-        }
-
-        // Update the focus-chain if necessary
-        // Always do this before sending focus change, since this event updates the focus chain.
-        if self.root_state().update_focus_chain {
-            run_update_focus_chain_pass(self);
-        }
-
-        run_update_focus_pass(self, widget_state);
-
-        if self.root_state().request_anim {
+    // --- MARK: REWRITE PASSES ---
+    pub(crate) fn run_rewrite_passes(&mut self) {
+        // TODO - Rerun passes if invalidation flags are still set
+
+        run_mutate_pass(self);
+        run_update_widget_tree_pass(self);
+        run_update_disabled_pass(self);
+        run_update_stashed_pass(self);
+        run_update_focus_chain_pass(self);
+        run_update_focus_pass(self);
+        root_layout(self);
```

Comment:
I think this runs a full layout pass even if it's not requested/needed doesn't it? (I haven't checked the other passes in detail, might be worth checking this as well)

Reply:
Oh, good catch.

Yeah, right now I think layout is the only pass with no "short circuits" whatsoever.

(Though update_pointer and update_focus could have more explicit short circuits too.)

Reply:
Added an early exist to root_layout.